icu4c/source/i18n/number_longnames.cpp

```diff
@@ -1410,13 +1405,15 @@ void MixedUnitLongNameHandler::forMeasureUnit(const Locale &loc,
                                               const MicroPropsGenerator *parent,
                                               MixedUnitLongNameHandler *fillIn,
                                               UErrorCode &status) {
+    U_ASSERT(mixedUnit.getComplexity(status) == UMEASURE_UNIT_MIXED);
```

Comment:
Since we don't get read-only access to MeasureUnitImpl in ICU4J, we can't do the quick/cheap check for impl.complexity there that we have below (line 1417). Instead, for ICU4J, I just trust the invariant, checked in debug mode via an assert.

Here in ICU4C, I've added the same assert, for some more consistency. Assert failures are anyway also much easier to debug in ICU4C than status-based failures.